### ISSUE TITLE
Add promise after finish save the file

### ISSUE THREAD

**Issue by NoelBroda** [jsierles/react-native-audio#115]
Hi man!
If I try to send the file to my server automatically I stop recording, the file size is `0`.
I have to wait for ~1 second, so the file finished to be saved in the `audioPath`, and then send it to my server. And in this case, it works.

```javascript
let audio = {
    uri: audioPath,
    type: 'audio/aac',
    name: 'audio.aac'
};

let body = new FormData();
body.append('audio', audio);

xhr.open('POST', 'http://xxxx.com/xxx.php');
xhr.send(body);
```

If I execute that code exactly after calling `AudioRecorder.stopRecording();`, It will not work.
I have to put that code in a function and call it using `setTimeout(xxx, 2000);`

Maybe, the `AudioRecorder.stopRecording()` function should return a promise, so I can do 
```javascript
AudioRecorder.stopRecording().then(function(){
    // here send the tile to my server
});
```

If you have no time to do it, let me know and I can try to add this behavior and send a PR.

**Comment by chrisijoyah** [comment]
In the AudioExample.js file where does it store the audio file on the device?

**Comment by jsierles** [comment]
Sorry, been busy and I'm not using this module. Feel free to submit a PR!

**Comment by NoelBroda** [comment]
I have no idea about Objective-C and Java-Android.
:shrug:

**Comment by NoelBroda** [comment]
I'm checking how you did it in another functions.
I'll be researching about it.
Maybe in a couple of days I can submit a PR.
I'll request some help in StackOverflow if I can do it by myself.

Thank you

**Comment by Sumrocks** [comment]
i have encountered this problem in Android,
i can't upload the file,because it's not saved!

**Comment by tomprogers** [comment]
@chrisijoyah The file is saved at the path you provide in your `AudioRecorder.prepareRecordingAtPath` call. This is demonstrated in the example:

- [here](https://github.com/jsierles/react-native-audio/blob/master/AudioExample/AudioExample.js#L82), where the example app decides what the path should be,
- and then [here](https://github.com/jsierles/react-native-audio/blob/master/AudioExample/AudioExample.js#L26), where the path is ultimately being provided to AudioRecorder's interface.

**Comment by jsierles** [comment]
Closing this for now in favor of a future PR!

**Comment by NoelBroda** [comment]
@tomprogers yes, I know where is saved, but, that have nothing related to "use promisses".

Let me give you this example:
* Call stop recording
* try to send the file to the server.

But, let's supouse the device is busy, and takes more than 500ms to save the file.
In the moment you are trying to send te file to the server, the file didn't exists because it was not saved yet.

This is why we need a promise what should be solved in the moment the file finished to be saved.
Meaning, I'm talking about "JS promises".

**Comment by NoelBroda** [comment]
```javascript
xxx.stopRecording().then(function() {
    // send the file to the server.
});
```

**Comment by jsierles** [comment]
I agree we should implement promises here. I'll look into this if nobody else gets to it first.

**Comment by NoelBroda** [comment]
I really tried it, but... zero knowledge in Objetive-C and Java :/

**Comment by NoelBroda** [comment]
Any news about it?

**Comment by cooperka** [comment]
@NoelBroda are you using the `AudioRecorder.onFinished` callback? `AudioRecorder.stopRecording` is asynchronous so you won't be able to access the file until you get `onFinished`. You can have a look at the [example app](https://github.com/jsierles/react-native-audio/blob/master/AudioExample/AudioExample.js) to see how it works.

**Comment by NoelBroda** [comment]
There is a difference between "when the record is finished" and "when the file is finished to be saved".
I'm simulating an old/slow device, and the `onFinished` event is executed to earlier before the file is totally saved in the device.

**Comment by cooperka** [comment]
@NoelBroda maybe Apple has a bug on iOS. Here's their documentation for [`audioRecorderDidFinishRecording`](https://developer.apple.com/reference/avfoundation/avaudiorecorderdelegate/1388688-audiorecorderdidfinishrecording), implemented [here](https://github.com/jsierles/react-native-audio/blob/master/ios/AudioRecorderManager.m#L77).

**Comment by jsierles** [comment]
I don't think we'll find any API for this. AVFoundation is simple but limited. The alternative here, as @NoelBroda suggests, could be to provide method in javascript that polls for the file size until it's a non-zero value, then returns. This isn't ideal, but would solve this problem. For now, you should be able to solve this by polling the `duration` value until it returns a non-zero value.

**Comment by NoelBroda** [comment]
Actually, that behavior could fail as well.
Something that should no fail is to check (in case it's possible) the file status. Is the file is being written, the file should be blocked by the SO until the process that is writing finish doing it.

But also you have another issue there. What if the native recorder first of all create the file (like a `touch` in linux) and then write it? Well, if we check the status between the touch and the moment start writing, we'll have another false positive.
Example, let's suppose the native recordar do something like this (pseudo code):
```
create_file('audio.xxx');
put_content_into_file('audio', theContent);
```
If you check the status exactly in the middle of those lines (we are talking about async problems), we'll have that false positive.

So what we should do, is something like this (pseudo-code):
```
while (file_size === 0) {
    // continue waiting until this file is being written
    sleep(10);
}
while (file_status === 'not_busy') {
    // continue waiting until it finished to be written
    // sleep(10);
}

// It finished!
onFinishedCallback()!
```

I know those `while()` are not the best for the performance, but I just wanted to show you a pseudo-code.

(sorry for my english)

**Comment by jsierles** [comment]
Yes, this sort of thing sounds reasonable. I don't have bandwidth to work on it - so hoping someone will tackle it!

**Comment by NoelBroda** [comment]
What about leaving this issue open, or something like that, so someone can tackle it?

**Comment by NoelBroda** [comment]
This issue should be taken into account.

**Comment by RakanNimer** [comment]
Hey @NoelBroda, 

I'm working on a fix for this, Will send you a link to test in a bit.

**Comment by RakanNimer** [comment]
I opened a PR with a fix, could you give it a try and let me know if it works as you'd expect ? 

The promise returns the filepath. I've found that to be useful when you don't necessarily care about persisting the path.

**Comment by NoelBroda** [comment]
Wow! That was quick!
Thank you a lot.
Next week I'll be in holidays (really far away from my home). I'll come back to my home April 16. I'll be testing it in the week that I'll come back to home!

Thank you a lot again!

**Comment by RakanNimer** [comment]
Ok cool have a nice vacation :) 

@cooperka and I will hopefully integrate it in a new version before you come back ! 

Cheers !
